### PR TITLE
P6 - get the locastore from swarm to bee

### PR DESCRIPTION
**Pull request by jmozah**
This PR contains the localstore from swarm
Refer P6 in #66

### CODE REVIEW

**Comment by jmozah** [comment]
@janos i fixed most of the lint errors anyway. but there are two of them that i think is right but linter is not getting satisfied. Please have a look and let me know if i am missing anything.

**Comment by janos** [comment]
@jmozah 

```
golangci-lint run
##[error]pkg/storage/localstore/subscription_pull_test.go:576:7: ineffectual assignment to `err` (ineffassign)
						err = fmt.Errorf("got chunk (bin id %v in bin %v) from retrieval index %s: %v", i, bin, addrs[bin][i].Hex(), err)
						^
##[error]pkg/storage/localstore/subscription_pull_test.go:579:8: ineffectual assignment to `err` (ineffassign)
							err = fmt.Errorf("got chunk bin id %v in bin %v %v, want %v", i, bin, got, want)
							^
##[error]Makefile:23: recipe for target 'lint' failed
```

These warnings are actually correct as errors are not handled, just assigned.

**Comment by acud** [comment]
@jmozah linter still failing

**Comment by jmozah** [comment]
@janos  Aren't those err channeled below ?

**Comment by janos** [comment]
> @janos Aren't those err channeled below in errChan?

That is exactly what is shadowed and linter complains about it.

File: pkg/storage/localstore/subscription_pull_test.go [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+			var err error
+			addrsMu.Lock()
+			if i+1 > len(addrs[bin]) {
+				err = fmt.Errorf("got more chunk addresses %v, then expected %v, for bin %v", i+1, len(addrs[bin]), bin)
+			} else {
+				addr := addrs[bin][i]
+				if !bytes.Equal(got.Address, addr) {
+					err = fmt.Errorf("got chunk bin id %v in bin %v %v, want %v", i, bin, got.Address.Hex(), addr.Hex())
+				} else {
+					want, err := db.retrievalDataIndex.Get(shed.Item{
```
Comment (acud) [inline pkg/storage/localstore/subscription_pull_test.go:572]:
```suggestion
					var want shed.Item
					want, err = db.retrievalDataIndex.Get(shed.Item{
```

Comment (acud) [reply pkg/storage/localstore/subscription_pull_test.go:572]: i think this should solve this

File: pkg/storage/localstore/subscription_pull_test.go [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+			addrsMu.Lock()
+			if i+1 > len(addrs[bin]) {
+				err = fmt.Errorf("got more chunk addresses %v, then expected %v, for bin %v", i+1, len(addrs[bin]), bin)
+			} else {
+				addr := addrs[bin][i]
+				if !bytes.Equal(got.Address, addr) {
+					err = fmt.Errorf("got chunk bin id %v in bin %v %v, want %v", i, bin, got.Address.Hex(), addr.Hex())
+				} else {
+					var want shed.Item
+					var err1 error
```
Comment (janos) [inline pkg/storage/localstore/subscription_pull_test.go:573]: No need to have err1 variable defined and then explicitly set to err. I think that err can be set in the line below, just as @acud suggested in previous comment.

File: go.mod [hunk 3]
```diff
@@ -1,33 +1,33 @@
...
+	github.com/spf13/viper v1.6.3
+	github.com/syndtr/goleveldb v1.0.0
 	github.com/uber/jaeger-client-go v2.22.1+incompatible
 	github.com/uber/jaeger-lib v2.2.0+incompatible // indirect
-	golang.org/x/crypto v0.0.0-20200311171314-f7b00557c8c4
-	golang.org/x/sync v0.0.0-20190911185100-cd5d95a43a6e
-	resenje.org/web v0.4.0
+	golang.org/x/crypto v0.0.0-20200406173513-056763e48d71
+	golang.org/x/sync v0.0.0-20200317015054-43a5402ce75a
+	resenje.org/web v0.4.3
```
Comment (janos) [inline go.mod:32]: Can we leave unrelated dependencies unchanged in this pr? To restore them to the ones from master. I don't think that we should mix updating all dependencies with adding a new package that do not depend on them.

**Comment by janos** [comment]
There are some strange build errors https://github.com/ethersphere/bee/pull/82/checks?check_run_id=580487321#step:9:9. CI is failing.

**Comment by jmozah** [comment]
> There are some strange build errors https://github.com/ethersphere/bee/pull/82/checks?check_run_id=580487321#step:9:9. CI is failing.

@janos Only mac build look like failing because of this error. Other are failing because of some test error which is unrelated to my changes.  Overall.. we have to fix few things in CI to make it reliable.

**Comment by acud** [comment]
@jmozah the mac build is failing then breaks the rest of the builds. I guess this has to do with the go mod revert.

**Comment by jmozah** [comment]
> Please fix the failure at build stage https://github.com/ethersphere/bee/pull/82/checks?check_run_id=580487321#step:9:9, so that we can do the review.

@janos i cannot access the link..  but it looks like what @acud said.. macos build fails and it stalls the other builds...  This error doesn't look like my error... help me fix this

```
Run make build
go build -trimpath -ldflags "-s -w" ./...
# github.com/elastic/gosigar
##[error]/Users/runner/go/pkg/mod/github.com/elastic/gosigar@v0.0.0-20180330100440-37f05ff46ffa/concrete_sigar.go:20:11: cpuUsage.Get undefined (type Cpu has no field or method Get)
##[error]/Users/runner/go/pkg/mod/github.com/elastic/gosigar@v0.0.0-20180330100440-37f05ff46ffa/concrete_sigar.go:30:13: cpuUsage.Get undefined (type Cpu has no field or method Get)
##[error]/Users/runner/go/pkg/mod/github.com/elastic/gosigar@v0.0.0-20180330100440-37f05ff46ffa/concrete_sigar.go:49:10: l.Get undefined (type LoadAverage has no field or method Get)
##[error]/Users/runner/go/pkg/mod/github.com/elastic/gosigar@v0.0.0-20180330100440-37f05ff46ffa/concrete_sigar.go:55:10: m.Get undefined (type Mem has no field or method Get)
##[error]/Users/runner/go/pkg/mod/github.com/elastic/gosigar@v0.0.0-20180330100440-37f05ff46ffa/concrete_sigar.go:61:10: s.Get undefined (type Swap has no field or method Get)
##[error]/Users/runner/go/pkg/mod/github.com/elastic/gosigar@v0.0.0-20180330100440-37f05ff46ffa/concrete_sigar.go:67:10: p.Get undefined (type HugeTLBPages has no field or method Get)
##[error]/Users/runner/go/pkg/mod/github.com/elastic/gosigar@v0.0.0-20180330100440-37f05ff46ffa/concrete_sigar.go:79:11: fd.Get undefined (type FDUsage has no field or method Get)
make: *** [build] Error 2
##[error]Process completed with exit code 2.
```

**Comment by janos** [comment]
@jmozah This error looks to me completely isolated to this PR.

I will try to fix it, but currently there are conflict in this PR. Could you solve them first?

**Comment by jmozah** [comment]
@janos  Fixed it

**Comment by janos** [comment]
Hm, it failed again https://github.com/ethersphere/bee/pull/82/checks?check_run_id=582223747#step:9:9. I will check it.

**Comment by jmozah** [comment]
> Hm, it failed again https://github.com/ethersphere/bee/pull/82/checks?check_run_id=582223747#step:9:9. I will check it.

@janos yeah, the same macos thingy..

**Comment by jmozah** [comment]
@janos one ubuntu build passed

**Pull request merged**